src/util/deps.ts

```diff
@@ -0,0 +1,52 @@
+import { INestApplicationContext } from "@nestjs/common"
+import { ModuleRef } from "@nestjs/core"
+import { useLogger } from "./logger"
+
+interface Type<T> extends Function {
+  new (...args: any[]): T;
+}
+type MaybePromise<T> = T | Promise<T> 
+
+let app: INestApplicationContext | undefined
+const buffering: Array<(ctx: INestApplicationContext) =>MaybePromise<unknown>> = []
+
+export function setGlobalApplicationContext(context: INestApplicationContext) {
+  if (app) {
+    throw new Error('ApplicationContext can not be set twice.')
+  }
+  app = context
+
+  while (buffering.length) {
+    const fn = buffering.shift()
+    if (fn) fn(context)
+  }
+}
+
+export function runWithDeps<RESULT>(
+  deps: Array<Type<unknown> | string | symbol>,
+  func: (...args: Array<string | symbol | any>) => MaybePromise<RESULT>,
+) {
+  const logger = useLogger('System')
+  const exec = (ctx: INestApplicationContext) => {
+    const moduleRef = ctx.get(ModuleRef)
+    return func(...deps.map((dep) => moduleRef.get(dep, { strict: false })))
+  }
+
+  if (app) return exec(app)
```

Comment:
This is interesting, could you explain what's the case that codes run after this line?

I saw the `app` is set from main.ts.

Reply:
First, when the application has been setup, use `setGlobalApplicationContext` to keep single context in it. Then u will have some situations to run some code with `runWithDeps`, this line means if `app` exists, exec the callback code immediately, or push it into `buffering` queue to exec

Reply:
just curious about what's the use case for `or push it into buffering queue to exec`,
as the `app` is set from the main.ts, it should exist, no?

Reply:
For design, the `runWithDeps` function used to collection todos and do callback if app ready, it will not limit to run with clearly `timeline`. We can use runWithDeps anywhere, but execute when app ready. In some actual case to consider app state is hard, but make a queue is easy.

Reply:
> For design, the `runWithDeps` function used to collection todos and do callback if app ready, it will not limit to run with clearly `timeline`. We can use runWithDeps anywhere, but execute when app ready. In some actual case to consider app state is hard, but make a queue is easy.

For this design, we can use `runWithDeps` in any lifecycle step of nestjs, such as `OnApplicatiionBootstrap`